pkg/clusterconfig/clusterconfig.go

```diff
@@ -102,21 +109,38 @@ func GetAWSConfig() (*Config, error) {
 	if err != nil {
 		return nil, err
 	}
+
 	cfg.Storage.Type = StorageTypeS3
 	if installConfig.Platform.AWS != nil {
 		cfg.Storage.S3.Region = installConfig.Platform.AWS.Region
 	}
-	sec, err := client.Secrets(installerConfigNamespace).Get(installerAWSCredsName, metav1.GetOptions{})
-	if err != nil {
-		return nil, fmt.Errorf("unable to read aws-creds secret: %v", err)
-	}
 
-	cfg.Storage.Type = StorageTypeS3
-	if v, ok := sec.Data["aws_access_key_id"]; ok {
-		cfg.Storage.S3.AccessKey = string(v)
-	}
-	if v, ok := sec.Data["aws_secret_access_key"]; ok {
-		cfg.Storage.S3.SecretKey = string(v)
+	// Look for a user defined secret to get the AWS credentials from first
+	sec, err := client.Secrets(imageRegistryNamespace).Get(userPrivateConfigName, metav1.GetOptions{})
+	if err != nil && errors.IsNotFound(err) {
+		glog.Infof("Optional user defined AWS credentials in secret \"%s/%s\" not found, ignoring.", imageRegistryNamespace, userPrivateConfigName)
+		// If no user defined secret is found, use the system one
+		sec, err = client.Secrets(installerConfigNamespace).Get(installerAWSCredsName, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("unable to get secret  \"%s/%s\": %v", installerConfigNamespace, installerAWSCredsName, err)
+		}
+		glog.Infof("Using cluster defined AWS credentials from secret \"%s/%s\"", installerConfigNamespace, installerAWSCredsName)
+		if v, ok := sec.Data["aws_access_key_id"]; ok {
+			cfg.Storage.S3.AccessKey = string(v)
+		}
+		if v, ok := sec.Data["aws_secret_access_key"]; ok {
+			cfg.Storage.S3.SecretKey = string(v)
+		}
+	} else if err != nil {
+		return nil, err
+	} else {
+		glog.Infof("Using user defined AWS credentials from secret \"%s/%s\"", imageRegistryNamespace, userPrivateConfigName)
```

Comment:
need to handle watching this secret in case the value is changed (or the secret is introduced later.  or deleted later).

Reply:
Should I introduce that code directly into https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/operator/controller.go#L162 which runs periodically , or should it be some kind of separate watcher?

Reply:
@coreydaley it handles events from watchers: https://github.com/openshift/cluster-image-registry-operator/blob/83f0b8d3eddf2175496f915b7c095baa1d7f57a9/pkg/operator/controller.go#L282-L292

Reply:
yeah it should be part of the existing controller which is already watching secrets anyway, you just need to add the logic to handle *this* secret.